File: ocaml/bin/p0036.ml
(* Project Euler: Problem 36 *)

open Core

let solve num =
  let is_palindrome = Euler.Math.is_palindrome in
  let rec aux n result =
    if n < 1 then
      result
    else
      if Bool.(is_palindrome n = true) && Bool.(is_palindrome n ~base:2 = true) then
        aux (n - 2) (result + n)
      else
        aux (n - 2) result
  in
  aux num 0

let exec () =
  Int.to_string (solve (1_000_000 - 1))

let () = Euler.Task.run exec

File: rust/src/bin/p0036.rs
// Project Euler: Problem 36

euler::run_solver!(36);

fn solve() -> String {
    compute(1_000_000).to_string()
}

fn compute(limit: u64) -> u64 {
    use euler::math;

    fn check_palindrome(n: u64) -> bool {
        math::is_palindrome(n, 10) && math::is_palindrome(n, 2)
    }

    debug_assert!(limit > 1);

    (1..limit).step_by(2).filter(|x| check_palindrome(*x)).sum()
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0036_limit_1000000() {
        assert_eq!(compute(1_000_000), 872187);
    }
}
